File: content/snippets/scala2/design_patterns/command.scala
// Command
class Device {
    def turnOn(): Unit = println("Device is ON")
    def turnOff(): Unit = println("Device is OFF")
}

trait Command {
    def execute(): Unit
}

class TurnOnCommand(device: Device) extends Command {
    def execute(): Unit = device.turnOn()
}

class TurnOffCommand(device: Device) extends Command {
    def execute(): Unit = device.turnOff()
}

class Invoker {
    private var commands: List[Command] = List()
    
    def addCommand(cmd: Command): Unit = {
        commands = commands :+ cmd
    }
    
    def executeAll(): Unit = {
        commands.foreach(_.execute())
    }
}

object Main extends App {
    val device = new Device()
    val invoker = new Invoker()
    
    invoker.addCommand(new TurnOnCommand(device))
    invoker.addCommand(new TurnOffCommand(device))
    
    invoker.executeAll()
}

File: content/snippets/go/design_patterns/command.go
package main

import "fmt"

type Receiver struct{}

func (r *Receiver) TurnOn() {
	fmt.Println("Device is ON")
}

func (r *Receiver) TurnOff() {
	fmt.Println("Device is OFF")
}

type Command interface {
	Execute()
}

type TurnOnCommand struct {
	receiver *Receiver
}

func (c *TurnOnCommand) Execute() {
	c.receiver.TurnOn()
}

type TurnOffCommand struct {
	receiver *Receiver
}

func (c *TurnOffCommand) Execute() {
	c.receiver.TurnOff()
}

type Invoker struct {
	commands []Command
}

func (i *Invoker) AddCommand(cmd Command) {
	i.commands = append(i.commands, cmd)
}

func (i *Invoker) ExecuteAll() {
	for _, cmd := range i.commands {
		cmd.Execute()
	}
}

func main() {
	device := &Receiver{}
	invoker := &Invoker{}

	invoker.AddCommand(&TurnOnCommand{device})
	invoker.AddCommand(&TurnOffCommand{device})

	invoker.ExecuteAll()
}
